migration/orm_test.go

```diff
@@ -0,0 +1,154 @@
+	const thisPkgName = "testpkg"
+
+	reg := newRegister()
+
+	reg.MustRegister(1, &MyModel{}, NoModification)
+	reg.MustRegister(2, &MyModel{}, func(db weave.ReadOnlyKVStore, m Migratable) error {
+		msg := m.(*MyModel)
+		msg.Cnt += 2
+		return msg.err
+	})
+	reg.MustRegister(3, &MyModel{}, func(db weave.ReadOnlyKVStore, m Migratable) error {
+		panic("not implemented")
+	})
+
+	db := store.MemStore()
+
+	schema := NewSchemaBucket()
+	if _, err := schema.Create(db, &Schema{Metadata: &weave.Metadata{Schema: 1}, Pkg: thisPkgName, Version: 1}); err != nil {
+		t.Fatalf("cannot register schema version: %s", err)
+	}
+
+	b := &MyModelBucket{
+		Bucket: NewBucket(thisPkgName, "mymodel", orm.NewSimpleObj(nil, &MyModel{})),
+	}
+
+	// Use custom register instead of the global one to avoid pollution
+	// from the application during tests.
+	b.Bucket.migrations = reg
+
+	obj := orm.NewSimpleObj([]byte("schema_one"), &MyModel{
+		Metadata: &weave.Metadata{Schema: 1},
+		Cnt:      5,
+	})
+	assert.Nil(t, b.Save(db, obj))
+
+	if m, err := b.GetMyModel(db, "schema_one"); err != nil {
+		t.Fatalf("cannot get model one: %s", err)
+	} else if m.Metadata.Schema != 1 || m.Cnt != 5 {
+		t.Fatalf("unexpected result model: %#v", m)
+	}
+
+	// Storing a model with a schema version higher than currently active
+	// is not allowed.
+	obj = orm.NewSimpleObj([]byte("schema_two"), &MyModel{
+		Metadata: &weave.Metadata{Schema: 2},
+		Cnt:      11,
+	})
+	if err := b.Save(db, obj); !errors.ErrSchema.Is(err) {
+		t.Fatalf("storing an object with an unknown schema version: %s", err)
+	}
+
+	// Bumping a schema should unlock saving entities with higher schema version.
+	if _, err := schema.Create(db, &Schema{Metadata: &weave.Metadata{Schema: 1}, Pkg: thisPkgName, Version: 2}); err != nil {
+		t.Fatalf("cannot register schema version: %s", err)
+	}
+
+	if err := b.Save(db, obj); err != nil {
+		t.Fatalf("cannot save second object after schema version update: %s", err)
+	}
+
+	// Now that the schema was upgraded, all returned modlels must use it.
+	// This means that returned models metadata schema must be set to two
+	// and the payload must be updated.
+
+	if m, err := b.GetMyModel(db, "schema_one"); err != nil {
+		t.Fatalf("cannot get first model: %s", err)
+	} else if m.Metadata.Schema != 2 || m.Cnt != 5+2 {
+		t.Fatalf("unexpected result model: %#v", m)
+	}
+
+	if m, err := b.GetMyModel(db, "schema_two"); err != nil {
+		t.Fatalf("cannot get second model: %s", err)
+	} else if m.Metadata.Schema != 2 || m.Cnt != 11 {
+		t.Fatalf("unexpected result model: %#v", m)
+	}
+
+	// Saving a model with an outdated schema must call the migration
+	// before writing to the database.
+	obj = orm.NewSimpleObj([]byte("schema_one_2"), &MyModel{
+		Metadata: &weave.Metadata{Schema: 1},
+		Cnt:      17,
+	})
+	assert.Nil(t, b.Save(db, obj))
+}
+
+type MyModelBucket struct {
+	Bucket
+}
+
+func (b *MyModelBucket) GetMyModel(db weave.KVStore, key string) (*MyModel, error) {
```

Comment:
We really need to get rid of this boilerplate one day...

Reply:
This is not hard to implement. It is problematic to change the API because it will require a lot of code to be changed. :see_no_evil: